arch/arm64/src/imx9/imx9_usdhc.c

```diff
@@ -0,0 +1,3364 @@
+{
+  unsigned int watermark;
+  union
+  {
+    uint32_t w;
+    uint8_t  b[4];
+  } data;
+
+  /* Set the Read Watermark Level to 1: BRR will be set when the number of
+   * queued words is greater than or equal to 1.
+   */
+
+  putreg32(1 << USDHC_WML_RD_SHIFT, priv->addr + IMX9_USDHC_WML_OFFSET);
+
+  /* Loop while there is space to store the data, waiting for buffer
+   * read ready (BRR)
+   */
+
+  mcinfo("Entry: remaining: %lu IRQSTAT: %08x\n", priv->remaining,
+         getreg32(priv->addr + IMX9_USDHC_IRQSTAT_OFFSET));
+
+  while (priv->remaining > 0 &&
+         (getreg32(priv->addr + IMX9_USDHC_IRQSTAT_OFFSET) &
+          USDHC_INT_BRR) != 0)
+    {
+      /* Clear BRR.  If there is more data in the buffer, reading from the
+       * buffer should reset BRR.
+       */
+
+      putreg32(USDHC_INT_BRR, priv->addr + IMX9_USDHC_IRQSTAT_OFFSET);
+
+      /* Read the next word from the RX buffer */
+
+      data.w = getreg32(priv->addr + IMX9_USDHC_DATAPORT_OFFSET);
+      if (priv->remaining >= sizeof(uint32_t))
+        {
+          /* Transfer the whole word to the user buffer */
+
+          *priv->buffer++  = data.w;
+          priv->remaining -= sizeof(uint32_t);
+        }
+      else
+        {
+          /* Transfer any trailing fractional word */
+
+          uint8_t *ptr = (uint8_t *)priv->buffer;
+          int i;
+
+          for (i = 0; i < priv->remaining; i++)
+            {
+              *ptr++ = data.b[i];
+            }
+
+          /* Now the transfer is finished */
+
+          priv->remaining = 0;
+        }
+    }
+
+  /* Set the Read Watermark Level either the number of remaining words to be
+   * read (limited to half of the maximum watermark value)
+   */
+
+  watermark = ((priv->remaining + 3) >> 2);
+  if (watermark > (USDHC_MAX_WATERMARK / 2))
+    {
+      watermark = (USDHC_MAX_WATERMARK / 2);
+    }
+
+  putreg32(watermark << USDHC_WML_RD_SHIFT,
+           priv->addr + IMX9_USDHC_WML_OFFSET);
+
+  mcinfo("Exit: remaining: %lu IRQSTAT: %08x WML: %08x\n", priv->remaining,
+         getreg32(priv->addr + IMX9_USDHC_IRQSTAT_OFFSET),
+         getreg32(priv->addr + IMX9_USDHC_WML_OFFSET));
+}
+#endif
+
+/****************************************************************************
+ * Name: imx9_recvdma
+ *
+ * Description:
+ *   Receive SDIO data in dma mode
+ *
+ * Input Parameters:
+ *   priv  - Instance of the SDMMC private state structure.
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#if defined(CONFIG_IMX9_USDHC_DMA) && !defined(CONFIG_ARM64_DCACHE_DISABLE)
+static void imx9_recvdma(struct imx9_dev_s *priv)
+{
+  unsigned int watermark;
+
+  if (priv->unaligned_rx)
+    {
+      /* If we are receiving multiple blocks to an unaligned buffers,
+       * we receive them one-by-one
+       */
+
+      /* Copy the received data to client buffer */
+
+      memcpy(priv->buffer, priv->rxbuffer, priv->blocksize);
```

Comment:
These operations are the wrong way around, cache must be invalidated first (to flush data from RAM), otherwise the CPU might just see the cache line representing priv->rxbuffer

Reply:
Thanks, this seems to resolve the problem.